pkg/spec/util.go

```diff
@@ -0,0 +1,214 @@
+	if err := isPointerToStruct(inputStruct); err != nil {
+		return errors.Wrap(err, "Input parameter type mismatch")
+	}
+
+	val := reflect.ValueOf(inputStruct).Elem()
+
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Type().Field(i)
+
+		// We do not need to store JSON tags like "inline" or "omitempty",
+		// instead we only need JSON tag which is used for unmarshalling the
+		// YAML. It's safe to assume that, that JSON tag is the one that
+		// is specified first.
+		// Let's split the JSON tag on ",", and store the first element of the
+		// resulting array.
+
+		unmarshalJsonTag := strings.SplitN(field.Tag.Get("json"), ",", 2)[0]
+
+		skipTag := false
+		// Populating only the non blacklisted tags
+		for _, blacklistedTag := range blacklistedTags {
+			if unmarshalJsonTag == blacklistedTag {
+				skipTag = true
+			}
+		}
+		if unmarshalJsonTag == "" {
+			// It does not make any sense to store anything for JSON tags with no
+			// first element, e.g. `json:",inline"`
+			skipTag = true
+		}
+
+		if skipTag {
+			continue
+		}
+		inputMap[unmarshalJsonTag] = append(inputMap[unmarshalJsonTag], val.Type().String())
+	}
+	return nil
+}
+
+// This function returns an array of pointers to all of the embedded structs
+// in the input struct.
+// The embedded structs in an embedded struct are also taken into considertation
+// in a recursive manner.
+func getEmbeddedStructs(inputStruct interface{}) ([]interface{}, error) {
+
+	// Checking if input is pointer to struct
+	if err := isPointerToStruct(inputStruct); err != nil {
+		return nil, errors.Wrap(err, "Input parameter type mismatch")
+	}
+
+	var embeddedStructs []interface{}
+
+	val := reflect.ValueOf(inputStruct).Elem()
+
+	for i := 0; i < val.NumField(); i++ {
+		// Checking for embedded structs
+		if val.Type().Field(i).Anonymous {
+
+			// Appending pointer to the embedded/anonymous struct to
+			// embeddedStructs
+			embeddedStructs = append(embeddedStructs, val.Field(i).Addr().Interface())
+
+			// Since the current field is an anonymous struct, we call the
+			// function recursively to get embedded structs under it
+			recursiveEmbeddedStructs, err := getEmbeddedStructs(val.Field(i).Addr().Interface())
+			if err != nil {
+				return nil, errors.Wrapf(err, "Unable to get embedded structs recursively from %v", val.Field(i).String())
+			}
+
+			// Adding the returned recursive structs to embeddedStructs
+			embeddedStructs = append(embeddedStructs, recursiveEmbeddedStructs...)
+		}
+	}
+	return embeddedStructs, nil
+}
+
+// This function returns an array of all the JSON tags used for unmarshalling
+// (which are the first ones specified) which have a JSON tag "conflicting", in
+// all of the input structs.
+// It's made sure that only the tags in "package spec" are checked and all other
+// packages are ignored. This is done because we are expecting the structs to
+// be handled, i.e. marked as "conflicting" only in spec.go, and we do not want
```

Comment:
spec.go; since that's where the input structs and the entire spec resides, and all of the other structs are from upstream Kubernetes.
Input structs being from spec.go, and `val.Type().PkgPath()` together make sure that only spec.go is being checked.